src/runtime/loader.ts

```diff
@@ -99,14 +189,16 @@ export function mergeRequiredAPIs(requiredAPIs: ManifestRequiredAPI[]): Manifest
 export async function loadStack(functionsDir: string): Promise<ManifestStack> {
   const endpoints: Record<string, ManifestEndpoint> = {};
   const requiredAPIs: ManifestRequiredAPI[] = [];
+  const extensions: Record<string, ManifestExtension> = {};
```

Comment:
can extensions just be undefined if we don't have any extensions?

that feels like a safer, backward compat change (just introduce new key whenever you are defining the new extensions functions)

Reply:
I tested to make sure that detectFromYaml and detectFromPort handle "extra" things by ignoring them so this change is backwards compatible as is. Technically, endpoints and extensions are at the same level. We could have extensions and no endpoints, endpoints and no extensions, or both, or neither. I'd like that to be reflected in the code. They should be handled the same way.

Reply:
Do we think defining just the extensions as a valid use case? Not questioning whether it's technically valid - I just didn't think it's a recommended or even endorsed way to manage and deploy extensions.

Shouldn't you be using `extensions` manifest in your firebase.json to do that?

And this addition in Functions SDK is mainly to provide a way to easily write custom event handler as CF3 functions to extend your extensions?

Reply:
Yes, it's a valid use case. This is not just for functions defined for extensions. It's the new way to define extensions apart from the manifest. It's useful for handling multiple similar extensions as well.
My original plan was not to involve functions code at all and have all the extensions definitions handled by extensions code, but Thomas and Michael insisted that the new extensions definitions MUST live in the functions codebase and MUST be parsed by the functions SDK. It made everything much more complicated because extensions can now be deployed using `firebase deploy --only functions` (so complexity x 10 as there are now multiple places to deploy extensions and they must not clobber each other etc.) Which I thought was super confusing. But Thomas insisted and said he would do "everything in his power" to prevent me from doing it any other way. So here we are :)